rust/src/treefile.rs

```diff
@@ -0,0 +1,191 @@
+/* Copied and adapted from:
+ * https://github.com/cgwalters/coreos-assembler
+ * */
+
+extern crate libc;
+
+#[macro_use]
+extern crate serde_derive;
+extern crate serde;
+extern crate serde_json;
+extern crate serde_yaml;
+extern crate tempfile;
+
+use std::fs;
+use std::io;
+use std::io::Write;
+use std::ffi::CStr;
+use std::ffi::OsStr;
+use std::path::Path;
+use std::os::unix::ffi::OsStrExt;
+use std::os::unix::io::FromRawFd;
+
+#[no_mangle]
+pub extern fn treefile_read(filename: *const libc::c_char, fd: libc::c_int) -> libc::c_int {
+
+    // using an O_TMPFILE is an easy way to avoid ownership transfer issues w/ returning allocated
+    // memory across the Rust/C boundary
+
+    // let's just get the unsafory stuff (see what I did there?) out of the way first
+    let output_file = unsafe { fs::File::from_raw_fd(libc::dup(fd)) };
+    let c_str: &CStr = unsafe { CStr::from_ptr(filename) };
+    let filename_path = Path::new(OsStr::from_bytes(c_str.to_bytes()));
+
+    match treefile_read_impl(filename_path, output_file) {
+        Err(e) => {
+            // just write it out to stderr for now
```

Comment:
A whole world I haven't looked at is the [glib-rs bits](http://gtk-rs.org/docs/glib/error/struct.Error.html).  If we were to do this more seriously we'd probably need to pull that in.

Reply:
I spent some time trying to figure out how to use this, but it seems like these bindings are more meant for Rust to interact with glib rather than Rust pretending to be glib. E.g. there's an `Error::wrap()` method for wrapping GErrors received from glib, but I don't know how to create an `Error::new()` and pass it back as an unowned `GError` to rpm-ostree.

Reply:
@jlebon I think you are looking for https://docs.rs/glib-sys/0.6.0/glib_sys/struct.GError.html, which can be directly built by consumers (I am not familiar with glib, so I don't know who is supposed to own `message`). That said I think you don't want to pull all of `glib-rs` in here (as it will likely create linkage issues with its `build.rs`), but only some ctypes definitions for exchange at FFI border.

Reply:
@lucab Thanks for the pointer. I finally got this working by just directly using the `glib-sys` binding for `g_error_new_literal`. This works around the main hurdle of somehow having to pass ownership of an allocated string on the Rust side to C to be freed by `free`, which AFAICT is tricky business. OTOH, `g_error_new_literal` just does an `strdup`.

I made this into a trait to make it extra easy to use! :)